barretenberg/rust/barretenberg-rs/src/backends/ffi.rs

```diff
@@ -0,0 +1,152 @@
+//! FFI backend for Barretenberg
+//!
+//! This backend calls the Barretenberg C API directly via FFI,
+//! eliminating process spawn overhead. Ideal for mobile and embedded use cases.
+//!
+//! # Requirements
+//!
+//! This backend requires linking against `libbarretenberg`. You must:
+//! 1. Build Barretenberg as a static library (`libbarretenberg.a`)
+//! 2. Set the library search path via `BARRETENBERG_LIB_DIR` or cargo config
+//!
+//! # Example
+//!
+//! ```ignore
+//! use barretenberg_rs::{BarretenbergApi, backends::FfiBackend};
+//!
+//! let backend = FfiBackend::new()?;
+//! let mut api = BarretenbergApi::new(backend);
+//!
+//! let response = api.blake2s(b"hello world")?;
+//! println!("Hash: {:?}", response.hash);
+//! ```
+
+use crate::backend::Backend;
+use crate::error::{BarretenbergError, Result};
+use std::ptr;
+
+// C API exported by Barretenberg
+// See: barretenberg/cpp/src/barretenberg/bbapi/c_bind.hpp
+#[link(name = "barretenberg")]
+extern "C" {
+    /// Execute a msgpack-encoded command and return msgpack-encoded response.
+    ///
+    /// # Safety
+    /// - `input_in` must point to valid memory of `input_len_in` bytes
+    /// - `output_out` and `output_len_out` must be valid pointers
+    /// - Caller must free `*output_out` using `libc::free`
+    fn bbapi(
+        input_in: *const u8,
+        input_len_in: usize,
+        output_out: *mut *mut u8,
+        output_len_out: *mut usize,
+    );
+}
+
+/// FFI backend that calls Barretenberg directly via C API.
+///
+/// This is the most performant backend option as it avoids process spawning
+/// and IPC overhead. However, it requires linking against `libbarretenberg`.
+///
+/// # Thread Safety
+///
+/// This backend is **not** thread-safe. Each thread should have its own
+/// `FfiBackend` instance, or access should be synchronized externally.
+pub struct FfiBackend {
+    _initialized: bool,
+}
+
+impl FfiBackend {
+    /// Create a new FFI backend.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if Barretenberg initialization fails.
+    pub fn new() -> Result<Self> {
+        // Future: Could add SRS initialization here if needed
+        // For now, Barretenberg initializes lazily on first use
+        Ok(Self { _initialized: true })
+    }
+}
+
+impl Backend for FfiBackend {
+    fn call(&mut self, input: &[u8]) -> Result<Vec<u8>> {
+        let mut output_ptr: *mut u8 = ptr::null_mut();
+        let mut output_len: usize = 0;
+
```

Comment:
Looks good overall, eager to see this have a CI test that actually links the libbarretenberg.a file :)